### DISCUSSION TITLE
grouping?

### DISCUSSION THREAD

**Discussion by aolko** [adieyal/sd-dynamic-prompts#67] [Q&A]
What's the correct way to use dynamic prompts to group multiple objects or characters?
Like what if i want to group `cat, blue fur` and `dog, red fur, wearing a golden collar`

**Comment by adieyal** [comment]
Could you clarify? Is {cat,dog} one group and {blue fur, red fur} another group? Are you expecting combinations of the two, ie cat blue fur, dog blue fur, cat red fur, dog red fur?

**Reply by aolko** [reply]
what i mean is ordinary AND/OR grouping, like
```
{and tags go here} {and here} and here
{{or ~ tags ~ go ~ here} {or ~ here}} ~ {or ~ here}
```
(gelbooru example)

```
( bride ~ wedding_dress ~ wedding_veil ~ veil ) ( white_gloves ~ gloves ~ elbow_gloves ~ fingerless_gloves ~ white_legwear )
```
(r34 example)

**Reply by aolko** [reply]
hello? @adieyal

**Reply by adieyal** [reply]
Have you tried this:
```
{bride|wedding_dress|wedding_veil|veil} {white_gloves|gloves|elbow_gloves|fingerless_gloves|white_legwear}
```

The `{...}` syntax represents ORs, i.e. it will choose one of the options, AND is just the literal string.

Does that answer your question?

**Reply by aolko** [reply]
> Have you tried this:
> 
> ```
> {bride|wedding_dress|wedding_veil|veil} {white_gloves|gloves|elbow_gloves|fingerless_gloves|white_legwear}
> ```
> 
> The `{...}` syntax represents ORs, i.e. it will choose one of the options, AND is just the literal string.
> 
> Does that answer your question?

no it doesn't, need those ANDs, my use case is combining character/object descriptions

**Reply by adieyal** [reply]
Sorry, I'm trying to figure out what you're looking for but I don't understand your examples. How would your gelbooru example differ from my proposed solution? Can you provide an example of what you would imagine the prompt would look like as well as what the generated prompts would be?

For example:

Prompt: My favourite shape is a {red|green|blue} {triangle|square|circle}
Generated prompts: 
- My favourite shape is a green circle
- My favourite shape is a blue triangle
- etc

**Reply by aolko** [reply]
You have a fundamental problem understanding **AND queries**,
in plain text you would write an **AND query** like
```
cat, blue fur AND dog, red fur, wearing a golden collar
```
however, in templating, i would go about it as such:
```
{cat, blue fur} {dog, red fur, wearing a golden collar}
```
or
```
{cat, blue fur && dog, red fur, wearing a golden collar}
```
or, might as well change how both **AND** and **OR** parse entirely, and treat `{}` as groups, perhaps, maybe even `&` is excessive and just two or more groups should be treated as **AND**
```
{cat, blue fur}~{dog, red fur, wearing a golden collar}
{cat, blue fur}&{dog, red fur, wearing a golden collar}
```
meaning, you would treat that as
```
cat+blue+fur OR dog+red+fur+wearing+a+golden+collar
cat+blue+fur AND dog+red+fur+wearing+a+golden+collar
```
respectively

**Reply by aolko** [reply]
please don't drop out

**Reply by aolko** [reply]
well, you did drop out

**Reply by aolko** [reply]
do you still care about grouping **at all**?
